lib/src/null_safety_analysis.dart

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:analyzer/dart/analysis/context_builder.dart';
+import 'package:analyzer/dart/analysis/context_locator.dart';
+import 'package:pub/src/source/path.dart';
+
+import 'package:pub_semver/pub_semver.dart';
+import 'package:path/path.dart' as path;
+
+import 'entrypoint.dart';
+import 'io.dart';
+import 'package_name.dart';
+import 'pubspec.dart';
+import 'solver.dart';
+import 'source/cached.dart';
+import 'system_cache.dart';
+
+enum NullSafetyCompliance {
+  /// This package and all dependencies opted into null safety.
+  compliant,
+
+  /// This package opted into null safety, but some file or dependency is not
+  /// opted in.
+  apiOnly,
+
+  /// This package did not opt-in to null safety yet.
+  notCompliant,
+
+  /// The resolution failed. Or some dart file in a dependency
+  /// doesn't parse.
+  analysisFailed,
+}
+
+class NullSafetyAnalysis {
+  final SystemCache _systemCache;
+  final Map<PackageId, Future<NullSafetyCompliance>>
+      _packageInternallyGoodCache = {};
```

Comment:
Please document what this is... It's really cool.

Reply:
Done